File: todo-lsp/src/archive.rs
//! §アーカイブ — Archive (Alt+A) moves selected all-gray top-level blocks
//! under the root-level `Archive:` heading; Unarchive (Alt+Shift+A) moves
//! all-gray blocks under `Archive:` back to the document end.

use crate::format::{format_document, split_lines};
use crate::line;

/// One structure line of the document. `units` is the relative indent
/// measurement; `level` is the SPEC.md インデントレベル (`units / 4`) that
/// decides block membership.
struct Entry {
    line_idx: usize,
    units: usize,
    level: usize,
}

/// Structure entries (non-blank lines) in document order.
fn entries(lines: &[String]) -> Vec<Entry> {
    lines
        .iter()
        .enumerate()
        .filter_map(|(i, l)| {
            let p = line::parse_line(l);
            p.is_structure().then_some(Entry {
                line_idx: i,
                units: p.units,
                level: p.level,
            })
        })
        .collect()
}

/// Group entries into top-level blocks: a level-0 entry (インデントレベル 0,
/// including 1-3 leading spaces) plus every deeper entry that follows it.
/// Lines before the first level-0 entry belong to no top-level block.
fn top_level_blocks(lines: &[String]) -> Vec<Vec<usize>> {
    let mut blocks: Vec<Vec<usize>> = Vec::new();
    for e in entries(lines) {
        if e.level == 0 {
            blocks.push(Vec::new());
        }
        if let Some(block) = blocks.last_mut() {
            block.push(e.line_idx);
        }
    }
    blocks
}

/// Whether every member line of a block is a 灰色行 (has `@done`,
/// `@cancelled` or `@hide`).
fn block_is_all_gray(lines: &[String], block: &[usize]) -> bool {
    !block.is_empty()
        && block
            .iter()
            .all(|&i| line::parse_line(&lines[i]).gray().is_some())
}

/// Archive the selected all-gray top-level blocks under the root-level
/// `Archive:` heading (created at the document end when absent).
pub fn archive(text: &str, selection: &[usize]) -> String {
    let lines: Vec<String> = split_lines(text).iter().map(|s| s.to_string()).collect();
    if lines.is_empty() {
        return text.to_string();
    }
    let sel: Vec<usize> = selection.to_vec();
    let blocks = top_level_blocks(&lines);

    // Eligible: all-gray blocks containing a selected line. The `Archive:`
    // heading itself is never gray, so its own block can never qualify.
    let moved: Vec<usize> = blocks
        .iter()
        .filter(|b| block_is_all_gray(&lines, b) && b.iter().any(|i| sel.contains(i)))
        .flatten()
        .copied()
        .collect();
    if moved.is_empty() {
        return text.to_string();
    }

    // Insertion point: the end of the root-level `Archive:` heading's block,
    // or the document end when the heading must be created.
    let archive_line = root_archive_line(&lines);
    let (insert_after, create_heading) = match archive_line {
        Some(h) => (block_end(&lines, h), false),
        None => (lines.len() - 1, true),
    };

    // Rebuild: drop the moved lines, shift them one level deeper under
    // Archive, and reinsert them (normalized) after the insertion point.
    let shifted: Vec<String> = moved
        .iter()
        .map(|&i| {
            let raw = &lines[i];
            let parts = line::parse_line(raw);
            let depth = relative_depth(&lines, i);
            format!(
                "{}{}",
                line::indent_for_level(depth + 1),
                parts.normalize_body(raw)
            )
        })
        .collect();
    let mut out: Vec<String> = Vec::new();
    for (i, l) in lines.iter().enumerate() {
        if moved.contains(&i) {
            continue;
        }
        out.push(l.clone());
        if !create_heading && i == insert_after {
            out.extend(shifted.iter().cloned());
        }
    }
    if create_heading {
        out.push("Archive:".to_string());
        out.extend(shifted);
    }
    // §フォーマット applies after Archive execution (the moved lines and
    // the document as a whole are left in the canonical form).
    format_document(&join(&out, text))
}

/// Move all-gray blocks directly under the root-level `Archive:` heading
/// (containing a selected line) back to the document end, dedented to the
/// top level. Without an `Archive:` heading the document is unchanged.
pub fn unarchive(text: &str, selection: &[usize]) -> String {
    let lines: Vec<String> = split_lines(text).iter().map(|s| s.to_string()).collect();
    let sel: Vec<usize> = selection.to_vec();

    let archive_line = root_archive_line(&lines);
    let Some(archive_idx) = archive_line else {
        return text.to_string();
    };
    let end = block_end(&lines, archive_idx);

    // Direct-child sub-blocks of the root `Archive:` block: entries exactly
    // one level deeper (level 1), each with its own descendants. Deeper-only
    // content below the heading belongs to no direct-child block.
    let mut sub_blocks: Vec<Vec<usize>> = Vec::new();
    for e in entries(&lines) {
        if e.line_idx <= archive_idx || e.line_idx > end {
            continue;
        }
        if e.level == 1 {
            sub_blocks.push(Vec::new());
        }
        if let Some(block) = sub_blocks.last_mut() {
            block.push(e.line_idx);
        }
    }

    let moved: Vec<usize> = sub_blocks
        .iter()
        .filter(|b| block_is_all_gray(&lines, b) && b.iter().any(|i| sel.contains(i)))
        .flatten()
        .copied()
        .collect();
    if moved.is_empty() {
        return text.to_string();
    }

    // Dedent each moved line back to the top level (relative depth within
    // the moved sub-block kept) and append at the document end.
    let head_depth = relative_depth(&lines, moved[0]);
    let appended: Vec<String> = moved
        .iter()
        .map(|&i| {
            let raw = &lines[i];
            let parts = line::parse_line(raw);
            let depth = relative_depth(&lines, i) - head_depth;
            format!(
                "{}{}",
                line::indent_for_level(depth),
                parts.normalize_body(raw)
            )
        })
        .collect();
    let mut out: Vec<String> = Vec::new();
    for (i, l) in lines.iter().enumerate() {
        if moved.contains(&i) {
            continue;
        }
        out.push(l.clone());
    }
    out.extend(appended);
    // §フォーマット applies after Unarchive execution.
    format_document(&join(&out, text))
}

fn root_archive_line(lines: &[String]) -> Option<usize> {
    lines.iter().position(|line_text| {
        let parts = line::parse_line(line_text);
        parts.units == 0 && parts.is_archive_heading(line_text)
    })
}

/// The last line index of the block headed by the line at `head_idx`:
/// the head plus following entries at a deeper インデントレベル.
fn block_end(lines: &[String], head_idx: usize) -> usize {
    let head_level = line::parse_line(&lines[head_idx]).level;
    let mut end = head_idx;
    for e in entries(lines) {
        if e.line_idx <= head_idx {
            continue;
        }
        if e.level > head_level {
            end = e.line_idx;
        } else {
            break;
        }
    }
    end
}

/// The 0-based depth of a line within its top-level block (0 for the block
/// head, 1 for its children, ...), derived from the level chain.
fn relative_depth(lines: &[String], idx: usize) -> usize {
    let level = line::parse_line(&lines[idx]).level;
    let mut depth = 0;
    let mut cur = level;
    for e in entries(lines).into_iter().rev() {
        if e.line_idx >= idx || e.level >= level {
            continue;
        }
        if e.level < cur {
            depth += 1;
            cur = e.level;
        }
    }
    depth
}

fn join(lines: &[String], original: &str) -> String {
    let mut out = lines.join("\n");
    if original.ends_with('\n') || !out.is_empty() {
        out.push('\n');
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    const DONE_DOC: &str = "\
Inbox:
    a @done
    b
Archive:
    old @done
";

    #[test]
    fn archive_moves_selected_gray_top_level_block() {
        // Only top-level blocks can be archived: `a @done` is level 0.
        let input = "Inbox:\n    b\na @done\n";
        let out = archive(input, &[2]);
        assert_eq!(out, "Inbox:\n    b\n\nArchive:\n    a @done\n");
    }

    #[test]
    fn archive_into_existing_archive_block_end() {
        let input = "Inbox:\n    b\nArchive:\n    old @done\nc @done\n";
        let out = archive(input, &[4]);
        assert_eq!(out, "Inbox:\n    b\n\nArchive:\n    old @done\n    c @done\n");
    }

    #[test]
    fn archive_requires_all_lines_gray() {
        // The Inbox block contains non-gray lines (`b`), so nothing moves
        // even though a selected line is gray.
        let out = archive(DONE_DOC, &[1, 2]);
        assert_eq!(out, DONE_DOC);
    }

    #[test]
    fn archive_requires_selection() {
        assert_eq!(archive(DONE_DOC, &[2]), DONE_DOC);
    }

    #[test]
    fn archive_creates_archive_heading_when_absent() {
        let out = archive("t\ndone task @cancelled\n", &[1]);
        assert_eq!(out, "t\n\nArchive:\n    done task @cancelled\n");
    }

    #[test]
    fn archive_moves_whole_block_with_descendants() {
        let input = "x @done\n    deep @hide\ny\n";
        let out = archive(input, &[0]);
        assert_eq!(out, "y\n\nArchive:\n    x @done\n        deep @hide\n");
    }

    #[test]
    fn archive_moves_multiple_blocks_in_document_order() {
        let input = "L:\n    keep\none @done\ntwo @cancelled\n";
        let out = archive(input, &[2, 3]);
        assert_eq!(
            out,
            "L:\n    keep\n\nArchive:\n    one @done\n    two @cancelled\n"
        );
    }

    #[test]
    fn archive_block_under_archive_is_not_top_level() {
        // A gray block already inside Archive cannot be re-archived.
        let out = archive(DONE_DOC, &[4]);
        assert_eq!(out, DONE_DOC);
    }

    #[test]
    fn archive_ignores_indented_archive_heading_as_destination() {
        let input = "    Archive:\n        old @done\ndone @done\n";
        let out = archive(input, &[2]);
        // The archive decision ignored the indented `Archive:` (a new root
        // heading was created); §フォーマット then normalizes the indented
        // heading to the top level.
        assert_eq!(
            out,
            "Archive:\n    old @done\n\nArchive:\n    done @done\n"
        );
    }

    #[test]
    fn unarchive_requires_root_level_archive_heading() {
        let input = "    Archive:\n        old @done\n";
        assert_eq!(unarchive(input, &[1]), input);
    }

    #[test]
    fn unarchive_moves_block_to_document_end() {
        let input = "\
Inbox:
    a
Archive:
    old @done
        note @hide
    other @done
";
        let out = unarchive(input, &[3]);
        assert_eq!(
            out,
            "Inbox:\n    a\n\nArchive:\n    other @done\n\nold @done\n    note @hide\n"
        );
    }

    #[test]
    fn unarchive_without_archive_heading_is_noop() {
        let input = "a @done\n";
        assert_eq!(unarchive(input, &[0]), input);
    }

    #[test]
    fn unarchive_requires_selection() {
        let input = "Archive:\n    old @done\n";
        assert_eq!(unarchive(input, &[]), input);
    }

    #[test]
    fn unarchive_requires_gray_block() {
        let input = "Archive:\n    active\n";
        assert_eq!(unarchive(input, &[1]), input);
    }

    #[test]
    fn partial_indent_line_is_its_own_top_level_block() {
        // 1-3 leading spaces are still インデントレベル 0, so `   b @done` is
        // a separate top-level block from `a @done` and archives alone.
        let out = archive("a @done\n   b @done\n", &[1]);
        assert_eq!(out, "a @done\n\nArchive:\n    b @done\n");
    }

    #[test]
    fn leading_indented_line_is_not_a_top_level_block() {
        // A document starting with an indented line has no top-level block
        // there, so the selection cannot be archived.
        let input = "    a @done\nplain\n";
        assert_eq!(archive(input, &[0]), input);
    }

    #[test]
    fn blank_line_separates_top_level_blocks() {
        let out = archive("a @done\n\nb @done\n", &[2]);
        assert_eq!(out, "a @done\n\nArchive:\n    b @done\n");
    }

    #[test]
    fn archive_formats_the_whole_document() {
        // §フォーマット applies after Archive: the untouched ` t` line is
        // normalized too.
        let out = archive("  x @done\n t\n", &[0]);
        assert_eq!(out, "t\n\nArchive:\n    x @done\n");
    }

    #[test]
    fn unarchive_ignores_deeper_only_content_under_archive() {
        // Only direct children (level 1) of the root `Archive:` heading are
        // unarchive targets; a deeper line with no level-1 block does not move.
        let input = "Archive:\n        deep @done\n";
        assert_eq!(unarchive(input, &[1]), input);
    }
}

File: packages/todo-core/src/archive.ts
import { formatDocument, joinLines, splitLines } from "./format.js";
import {
  grayOf,
  indentForLevel,
  isArchiveHeading,
  isBlank,
  normalizeBody,
  parseLine,
} from "./line.js";

interface Entry {
  line: number;
  units: number;
  /** Index into the entry list of the 親行 — the nearest earlier non-blank
   * row of shallower indent. Undefined for ルート直下 rows (SPEC §構文要素). */
  parent: number | undefined;
}

/** Archive selected all-gray top-level blocks. */
export function archive(source: string, selection: readonly number[], tabSize = 4): string {
  const lines = splitLines(source);
  if (lines.length === 0) return source;

  const blocks = topLevelBlocks(lines, tabSize);
  const archiveLine = rootArchiveLine(lines, tabSize);
  const selected = new Set(selection);
  const moved = blocks
    .filter(
      (block) =>
        block.length > 0 &&
        !block.includes(archiveLine ?? -1) &&
        blockIsAllGray(lines, block) &&
        block.some((line) => selected.has(line)),
    )
    .flat();
  if (moved.length === 0) return source;

  const movedLines = moved.map((line) => shiftLine(lines, line, 1, tabSize));
  const output: string[] = [];
  if (archiveLine === undefined) {
    for (let line = 0; line < lines.length; line++) {
      if (!moved.includes(line)) output.push(lines[line]);
    }
    output.push("Archive:", ...movedLines);
  } else {
    const insertAfter = blockEnd(lines, archiveLine, tabSize);
    for (let line = 0; line < lines.length; line++) {
      if (moved.includes(line)) continue;
      output.push(lines[line]);
      if (line === insertAfter) output.push(...movedLines);
    }
  }
  return formatDocument(joinLines(output, source), tabSize);
}

/** Unarchive selected all-gray direct children of the root Archive heading. */
export function unarchive(source: string, selection: readonly number[], tabSize = 4): string {
  const lines = splitLines(source);
  const archiveLine = rootArchiveLine(lines, tabSize);
  if (archiveLine === undefined) return source;

  const end = blockEnd(lines, archiveLine, tabSize);
  const blocks = directChildBlocks(lines, archiveLine, end, tabSize);
  const selected = new Set(selection);
  const movedBlocks = blocks.filter(
    (block) => blockIsAllGray(lines, block) && block.some((line) => selected.has(line)),
  );
  const moved = movedBlocks.flat();
  if (moved.length === 0) return source;

  const output = lines.filter((_, line) => !moved.includes(line));
  for (const block of movedBlocks) {
    const rootDepth = relativeDepth(lines, block[0], tabSize);
    output.push(...block.map((line) => shiftLine(lines, line, -rootDepth, tabSize)));
  }
  return formatDocument(joinLines(output, source), tabSize);
}

/** トップレベルブロック: blocks headed by ルート直下 rows — rows with no
 * earlier non-blank row of shallower indent. */
function topLevelBlocks(lines: readonly string[], tabSize: number): number[][] {
  const blocks: number[][] = [];
  for (const entry of entries(lines, tabSize)) {
    if (entry.parent === undefined) blocks.push([]);
    blocks.at(-1)?.push(entry.line);
  }
  return blocks;
}

function directChildBlocks(
  lines: readonly string[],
  head: number,
  end: number,
  tabSize: number,
): number[][] {
  const structure = entries(lines, tabSize);
  const headIndex = structure.findIndex((entry) => entry.line === head);
  const blocks: number[][] = [];
  for (const entry of structure) {
    if (entry.line <= head || entry.line > end) continue;
    if (entry.parent === headIndex) blocks.push([]);
    blocks.at(-1)?.push(entry.line);
  }
  return blocks;
}

/** Non-blank rows with their indent units and parent index. */
function entries(lines: readonly string[], tabSize: number): Entry[] {
  const out: Entry[] = [];
  for (const [line, text] of lines.entries()) {
    const parts = parseLine(text, tabSize);
    if (isBlank(parts)) continue;
    let parent: number | undefined;
    for (let index = out.length - 1; index >= 0; index--) {
      if (out[index].units < parts.units) {
        parent = index;
        break;
      }
    }
    out.push({ line, units: parts.units, parent });
  }
  return out;
}

function blockIsAllGray(lines: readonly string[], block: readonly number[]): boolean {
  return block.length > 0 && block.every((line) => grayOf(parseLine(lines[line])) !== undefined);
}

/** The first `Archive:` 見出し行 at ルート直下 (an earlier row of shallower
 * indent would make it a child, not the root archive). */
function rootArchiveLine(lines: readonly string[], tabSize: number): number | undefined {
  for (const entry of entries(lines, tabSize)) {
    if (entry.parent !== undefined) continue;
    const line = lines[entry.line];
    if (isArchiveHeading(parseLine(line), line)) return entry.line;
  }
  return undefined;
}

/** The last row of `head`'s block: `head` plus its following deeper rows. */
function blockEnd(lines: readonly string[], head: number, tabSize: number): number {
  const structure = entries(lines, tabSize);
  const headUnits = structure.find((entry) => entry.line === head)?.units;
  let end = head;
  if (headUnits === undefined) return end;
  for (const entry of structure) {
    if (entry.line <= head) continue;
    if (entry.units > headUnits) end = entry.line;
    else break;
  }
  return end;
}

/** The depth of `target` in its parent chain (ルート直下 = 0). */
function relativeDepth(lines: readonly string[], target: number, tabSize: number): number {
  const structure = entries(lines, tabSize);
  let index = structure.findIndex((entry) => entry.line === target);
  let depth = 0;
  while (index >= 0 && structure[index].parent !== undefined) {
    index = structure[index].parent!;
    depth++;
  }
  return depth;
}

function shiftLine(
  lines: readonly string[],
  index: number,
  delta: number,
  tabSize: number,
): string {
  const line = lines[index];
  const parts = parseLine(line);
  const level = Math.max(0, relativeDepth(lines, index, tabSize) + delta);
  return `${indentForLevel(level, tabSize)}${normalizeBody(parts, line)}`;
}
